6 Sprint Commit

AlexSunset opened the pull request.

Commit:
6 Sprint Commit

pomegranate282922 commented on src/main/java/com/example/Lion.java:
```diff
@@ -5,18 +5,19 @@
+    Feline feline;
 
-    public Lion(String sex) throws Exception {
+    public Lion(String sex, Feline feline) throws Exception {
```
Отлично реализована изоляция.

pomegranate282922 commented on src/test/java/CatTest.java:
```diff
@@ -0,0 +1,31 @@
+public class CatTest {
+
+    @Mock
+    Feline feline = new Feline();
```
Мок строится на основе класса. Инициализировать объект не требуется. Нужно исправить.

pomegranate282922 commented on src/test/java/CatTest.java:
```diff
@@ -0,0 +1,31 @@
+        assertEquals("Мяу", cat.getSound());
+    }
+
+    @Test
```
Отлично! Уместное использование мока.

pomegranate282922 commented on src/test/java/FelineTest.java:
```diff
@@ -0,0 +1,37 @@
+    public void eatMeatCallsGetFood() throws Exception {
+        feline.eatMeat();
+        Mockito.verify(feline, Mockito.times(1)).getFood("Хищник");
+    }
```
Отлично! Уместное использование шпиона.

pomegranate282922 commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,67 @@
+            Assert.assertEquals("Используйте допустимые значения пола животного - самец или самка",
+                    e.getMessage());
+        }
+    }
```
Каждый тест этого класса повторится дважды, так как принадлежит параметризованному классу. Нужно разнести по разным классам.

pomegranate282922 commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Test
+    public void correctExceptionTextOnLionObjectCreate(){
+        try{
```
В такой реализации если метод не выкинет исключение, тест будет зелёным. 
нужно использовать метод assertThrows или параметр в аннотации Test.
Нужно исправить

pomegranate282922 reviewed (commented):
Работа хорошая! Изоляция выполнена верно, моки, стабы, шпионы и параметризация применены уместно.
Есть замечания:
1. Повторяющиеся одинаковые тесты в параметризованном классе 
2. Тест на исключение нужно реализовать по-другому 
3. Нет отчета jacoco

Commit:
- Убрал инициализацию объекта в CatTest
- Создал отдельный класс LionParametersTest для проверки конструктаора с различными параметрами
- Сделал дополнительный метод exceptionAppearOnLionObjectCreate для проверки самого факта появления исключения. Не смог реализовать в одном тесте и проверку на само исключение и на корректность выбрасываемого сообщения. Если так можно, прошу подсказать как
- Прикрепил отчёт в виде jacoco.exec. Если нужен другой способ предоставления отчёта, прошу подсказать)
-Спасибо за ревью

pomegranate282922 commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Test
+    public void exceptionAppearOnLionObjectCreate() {
+        Assert.assertThrows(Exception.class, () -> new Lion("Самец", feline));
```
Разве должно быть исключение на Самце?

AlexSunset replied:
Опечатался, да

pomegranate282922 commented on src/test/java/LionTest.java:
```diff
@@ -0,0 +1,50 @@
+            new Lion("Муфаса", feline);
+        } catch (Exception e) {
+            Assert.assertEquals("Используйте допустимые значения пола животного - самец или самка",
+                    e.getMessage());
```
В данной реализации если метод НЕ выкинет исключение, то тест останется зелёным. Нужно использовать метод assertThrows.

AlexSunset replied:
метод assertThrows проверяет только появление выбранного исключения, он не сверяет сам текст сообщения с нужным. Поэтому я создал два разных теста. В спецификации assertThrows не нашёл сигнатуры, которая бы проверяла ещё и сообщение - там есть возможность указать сообщение, но оно служит подсказкой при падении теста, а не сверяет текст сообщения при исключении с ожидаемым.
В методе явно указан исключительный случай - "Муфаса", в каком случае метод может не выкинуть исключение? Если вдруг в дальнейшем "Муфаса" станет валидным значением? Можно подробнее об этом?

pomegranate282922 reviewed (commented):
Многое поправил - супер! Но есть вопросики к тесту  на исключение. Нужно ещё раз посмотреть это место.
И прошу приложить всю папку jacoco из target.

AlexSunset commented:
Динара, добрый день!
У меня появилось пара вопросов, оставил их в комментариях к пулл реквесту
https://github.com/AlexSunset/qa_java/pull/5
Не могу закончить работу без из прояснения. Посмотри, если это возможно.
Спасибо!

вт, 4 апр. 2023 г. в 12:48, Dinara Trifanova ***@***.***>:

> ***@***.**** commented on this pull request.
>
> Многое поправил - супер! Но есть вопросики к тесту на исключение. Нужно
> ещё раз посмотреть это место.
> И прошу приложить всю папку jacoco из target.
> ------------------------------
>
> In src/test/java/LionTest.java
> <https://github.com/AlexSunset/qa_java/pull/5#discussion_r1156996036>:
>
> > +    public void getKittensCallsFeline() throws Exception {
> +        Lion lion = new Lion("Самец", feline);
> +        lion.getKittens();
> +        Mockito.verify(feline, Mockito.times(1)).getKittens();
> +    }
> +
> +    @Test
> +    public void eatMeatCallsGetFood() throws Exception {
> +        Lion lion = new Lion("Самец", feline);
> +        lion.getFood();
> +        Mockito.verify(feline, Mockito.times(1)).getFood("Хищник");
> +    }
> +
> +    @Test
> +    public void exceptionAppearOnLionObjectCreate() {
> +        Assert.assertThrows(Exception.class, () -> new Lion("Самец", feline));
>
> Разве должно быть исключение на Самце?
> ------------------------------
>
> In src/test/java/LionTest.java
> <https://github.com/AlexSunset/qa_java/pull/5#discussion_r1156999156>:
>
> > +        lion.getFood();
> +        Mockito.verify(feline, Mockito.times(1)).getFood("Хищник");
> +    }
> +
> +    @Test
> +    public void exceptionAppearOnLionObjectCreate() {
> +        Assert.assertThrows(Exception.class, () -> new Lion("Самец", feline));
> +    }
> +
> +    @Test
> +    public void correctExceptionTextOnLionObjectCreate() {
> +        try {
> +            new Lion("Муфаса", feline);
> +        } catch (Exception e) {
> +            Assert.assertEquals("Используйте допустимые значения пола животного - самец или самка",
> +                    e.getMessage());
>
> В данной реализации если метод НЕ выкинет исключение, то тест останется
> зелёным. Нужно использовать метод assertThrows.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AlexSunset/qa_java/pull/5#pullrequestreview-1370593943>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOA73K6XSJ6KP77ZZBLFB4LW7PVARANCNFSM6AAAAAAWQOLFXU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
- Переделал метод correctExceptionOnLionObjectCreate
- Добавил папку jacoco

pomegranate282922 reviewed (approved):
Да, теперь все ок!
